### PR TITLE
safe[] should support void returning functions

### PR DESCRIPTION
**Pull request by bkietz**
Also inlined `unwind_protect_unwrap_*`

### CODE REVIEW

File: inst/include/cpp11/protect.hpp [hunk 1]
```diff
@@ -159,8 +149,12 @@ void unwind_protect(Fun code) {
     throw unwind_exception(token);
   }
 
-  R_UnwindProtect(&internal::unwind_protect_unwrap_void<Fun>, &code, internal::maybe_jump,
-                  &unwind_data, token);
+  (void)R_UnwindProtect(
+      [](void* data) -> SEXP {
+        Fun* callback = (Fun*)data;
+        return R_NilValue;
```
Comment (jimhester) [inline inst/include/cpp11/protect.hpp:155]: I think you are missing the call to callback here right?

```
(*callback)();
```

File: inst/include/cpp11/protect.hpp [hunk 2]
```diff
@@ -179,42 +173,50 @@ void unwind_protect(Fun code) {
 struct protect {
   template <typename F>
   struct function;
-  template <typename R, typename... A>
-  struct function<R(A...)> {
-    constexpr R operator()(A... a) const {
+
+  template <typename... A>
+  struct function<void(A...)> {
+    constexpr void operator()(A... a) const {
```
Comment (jimhester) [inline inst/include/cpp11/protect.hpp:179]: At least with `-std=c++11` I am getting `Constexpr function's return type 'void' is not a literal type` for this. I think void constexpr are a C++14 feature.

**Comment by jimhester** [comment]
Thanks for working on this, unfortunately it doesn't seem yet able to compile cpp11test with these changes (see https://github.com/r-lib/cpp11/runs/903942342#step:10:194)

**Comment by jimhester** [comment]
With the following diff it gets closer, the void function `safe[R_CheckUserInterrupt]()` works. But the method resolution still fails on the C style variadic functions.

```diff
diff --git a/inst/include/cpp11/protect.hpp b/inst/include/cpp11/protect.hpp
index fb22b23..871f69b 100644
--- a/inst/include/cpp11/protect.hpp
+++ b/inst/include/cpp11/protect.hpp
@@ -152,6 +152,7 @@ void unwind_protect(Fun code) {
   (void)R_UnwindProtect(
       [](void* data) -> SEXP {
         Fun* callback = (Fun*)data;
+        (*callback)();
         return R_NilValue;
       },
       &code, internal::maybe_jump, &unwind_data, token);
@@ -176,7 +177,7 @@ struct protect {
 
   template <typename... A>
   struct function<void(A...)> {
-    constexpr void operator()(A... a) const {
+    void operator()(A... a) const {
       unwind_protect([&] { ptr_(a...); });
     }
     void (*ptr_)(A...);
@@ -184,18 +185,18 @@ struct protect {
 
   template <typename... A>
   struct function<SEXP(A...)> {
-    constexpr SEXP operator()(A... a) const {
+    SEXP operator()(A... a) const {
       return unwind_protect_sexp([&] { return ptr_(a...); });
     }
     SEXP (*ptr_)(A...);
   };
 
   template <typename R, typename... A>
-  constexpr function<R(A...)> operator[](R (&raw)(A...)) const {
+  function<R(A...)> operator[](R (&raw)(A...)) {
     return {&raw};
   }
 };
-constexpr struct protect safe = {};
+struct protect safe = {};
 
 inline void check_user_interrupt() { safe[R_CheckUserInterrupt](); }
 
```

```
  ../../inst/include/cpp11/protect.hpp:205:7: error: no viable overloaded operator[] for type 'struct protect'
     safe[Rf_error](fmt, args...);
     ~~~~^~~~~~~~~
   ../../inst/include/cpp11/protect.hpp:195:21: note: candidate template ignored: could not match 'R (A...)' against 'void (const char *, ...) __attribute__((noreturn))'
     function<R(A...)> operator[](R (&raw)(A...)) {
                       ^
   ../../inst/include/cpp11/protect.hpp:210:7: error: no viable overloaded operator[] for type 'struct protect'
     safe[Rf_error](fmt.c_str(), args...);
     ~~~~^~~~~~~~~
   ../../inst/include/cpp11/protect.hpp:195:21: note: candidate template ignored: could not match 'R (A...)' against 'void (const char *, ...) __attribute__((noreturn))'
     function<R(A...)> operator[](R (&raw)(A...)) {
                       ^
   ../../inst/include/cpp11/protect.hpp:215:7: error: no viable overloaded operator[] for type 'struct protect'
     safe[Rf_warning](fmt, args...);
     ~~~~^~~~~~~~~~~
   ../../inst/include/cpp11/protect.hpp:195:21: note: candidate template ignored: could not match 'R (A...)' against 'void (const char *, ...)'
     function<R(A...)> operator[](R (&raw)(A...)) {
                       ^
   ../../inst/include/cpp11/protect.hpp:220:7: error: no viable overloaded operator[] for type 'struct protect'
     safe[Rf_warning](fmt.c_str(), args...);
     ~~~~^~~~~~~~~~~
   ../../inst/include/cpp11/protect.hpp:195:21: note: candidate template ignored: could not match 'R (A...)' against 'void (const char *, ...)'
     function<R(A...)> operator[](R (&raw)(A...)) {
                       ^
```

File: inst/include/cpp11/protect.hpp (hunk 1 above)
Comment (bkietz) [reply inst/include/cpp11/protect.hpp:155]: yep, oops

File: inst/include/cpp11/protect.hpp [hunk 3]
```diff
@@ -144,13 +131,23 @@ SEXP unwind_protect_sexp(Fun code) {
...
+        return (*callback)();
+      },
+      &code, internal::maybe_jump, &unwind_data, token);
 }
 
-template <typename Fun>
+template <typename Fun, typename = typename std::enable_if<
+                            std::is_same<decltype(code()), SEXP>::value>::type>
+SEXP unwind_protect(Fun code) {
+  return unwind_protect_sexp(code);
```
Comment (bkietz) [inline inst/include/cpp11/protect.hpp:146]: @jimhester not sure if you'd prefer to remove or keep the explicitly suffixed overload

Comment (jimhester) [reply inst/include/cpp11/protect.hpp:146]: If the general case below handles this case I think we can remove it.

**Comment by jimhester** [comment]
Thanks for working on this!

**Pull request merged**